plugins/inputs/docker_log/docker_logs.go

```diff
@@ -0,0 +1,312 @@
+	filter, err := filter.NewIncludeExcludeFilter(d.LabelInclude, d.LabelExclude)
+	if err != nil {
+		return err
+	}
+	d.labelFilter = filter
+	return nil
+}
+
+func (d *DockerLogs) createContainerStateFilters() error {
+	if len(d.ContainerStateInclude) == 0 && len(d.ContainerStateExclude) == 0 {
+		d.ContainerStateInclude = []string{"running"}
+	}
+	filter, err := filter.NewIncludeExcludeFilter(d.ContainerStateInclude, d.ContainerStateExclude)
+	if err != nil {
+		return err
+	}
+	d.stateFilter = filter
+	return nil
+}
+
+func (d *DockerLogs) addToContainerList(containerId string, logReader io.ReadCloser) error {
+	d.mu.Lock()
+	defer d.mu.Unlock()
+	d.containerList[containerId] = logReader
+	return nil
+}
+
+func (d *DockerLogs) removeFromContainerList(containerId string) error {
+	delete(d.containerList, containerId)
+	return nil
+}
+
+func (d *DockerLogs) containerInContainerList(containerId string) bool {
+	if _, ok := d.containerList[containerId]; ok {
+		return true
+	}
+	return false
+}
+
+func (d *DockerLogs) stopAllReaders() error {
+	for _, container := range d.containerList {
+		container.Close()
+	}
+	return nil
+}
+
+func (d *DockerLogs) containerListUpdate(acc telegraf.Accumulator) error {
+	ctx, cancel := context.WithTimeout(context.Background(), d.Timeout.Duration)
+	defer cancel()
+	if d.client == nil {
+		log.Println("E! Error Dock client is null")
+		return nil
+	}
+	containers, err := d.client.ContainerList(ctx, d.opts)
+	if err != nil {
+		return err
+	}
+	for _, container := range containers {
+		if d.containerInContainerList(container.ID) {
+			continue
+		}
+		d.wg.Add(1)
+		/*Start a new goroutine for every new container that has logs to collect*/
+		go func(c types.Container) {
+			defer d.wg.Done()
+			err := d.getContainerLogs(c, acc)
+			if err != nil {
+				d.removeFromContainerList(c.ID)
+				log.Println(err)
+				return
+			}
+		}(container)
+	}
+	return nil
+}
+
+func (d *DockerLogs) getContainerLogs(
+	container types.Container,
+	acc telegraf.Accumulator,
+) error {
+	logOptions := types.ContainerLogsOptions{
+		ShowStdout: true,
+		ShowStderr: true,
+		Timestamps: false,
+		Details:    true,
+		Follow:     true,
+		Tail:       "0",
+	}
+	logReader, err := d.client.ContainerLogs(context.Background(), container.ID, logOptions)
+	d.addToContainerList(container.ID, logReader)
+	if err != nil {
+		log.Printf("Error getting docker logs: %s", err.Error())
+		return err
+	}
+	/* Parse container name */
+	cname := "unknown"
+	if len(container.Names) > 0 {
+		/*Pick first container name*/
+		cname = strings.TrimPrefix(container.Names[0], "/")
+	}
+
+	tags := map[string]string{
+		"containerId":   container.ID,
+		"containerName": cname,
+	}
+	// Add labels to tags
+	for k, label := range container.Labels {
+		if d.labelFilter.Match(k) {
+			tags[k] = label
+		}
+	}
+	fields := map[string]interface{}{}
+	data := make([]byte, LOG_BYTES_MAX)
+	for {
+		num, err := logReader.Read(data)
```

Comment:
Just realized that lines longer than `LOG_BYTES_MAX` will not have this prefix. I guess we should be reading that log header and consuming the length sent..

They use [`github.com/docker/docker/stdcopy`](https://github.com/moby/moby/blob/master/pkg/stdcopy/stdcopy.go) to do so. Perhaps you could look at using that or implementing a similar solution?

Reply:
Thanks for pointing this out ! I took most of the code from stdcopy and ensured that we're writing to the telegraf Accumulator instead of an io.Writer . Let me know if this approach works!

Reply:
Also I'm not sure why CI/CD is failing . I'm getting errors at some other location . Would you be able to help me out with this?

Reply:
I've been testing this for a while and still get a space before the log message (different behavior than docker cli) I wonder if it needs to be:
```
fields["log"] = buf[stdWriterPrefixLen+1 : frameSize+stdWriterPrefixLen]
```

Another issue I found is when running a docker container with `-t`, the stdcopy isn't valid and produces:
```
E! [inputs.docker_log] : Unrecognized input header: xx
```

The cli selectively uses stdcopy to strip the header [as seen here](https://github.com/docker/cli/blob/19.03/cli/command/container/logs.go#L74)

Reply:
Regarding the tests, I believe if you merge master into your branch, they should succeed. Also, please refrain from force pushing as there is an ongoing review :wink:

Reply:
> I've been testing this for a while and still get a space before the log message (different behavior than docker cli) I wonder if it needs to be:
> 
> ```
> fields["log"] = buf[stdWriterPrefixLen+1 : frameSize+stdWriterPrefixLen]
> ```
> Another issue I found is when running a docker container with `-t`, the stdcopy isn't valid and produces:
> 
> ```
> E! [inputs.docker_log] : Unrecognized input header: xx
> ```
> The cli selectively uses stdcopy to strip the header [as seen here](https://github.com/docker/cli/blob/19.03/cli/command/container/logs.go#L74)

Fixed the -t issue . I added a pushTtyLogs function based on what containerInspect returns.I noticed that the extra space in the front is also present in Tty mode . I think in both cases after we read , we probably should skip the first character of the string read from the logReader . Although I'm not sure if thats the right approach.

Reply:
I believe the first byte in the logreader is used to determine the log type, so dropping it wouldn't be ideal. It seems like it would be ok to skip one more byte on `buf` before assigning it to `fields["log"]`, though we should double check it works well with extra long logs

Reply:
I fixed this in the latest commit . I tested this as well and it seems to be working fine . Let me know if this is okay!